File: src/com/chbachman/cminus/representation/control/ForStatement.java
package com.chbachman.cminus.representation.control;

import com.chbachman.cminus.CMinusParser;
import com.chbachman.cminus.representation.Scope;
import com.chbachman.cminus.representation.Type;
import com.chbachman.cminus.representation.statement.Statement;
import com.chbachman.cminus.representation.value.Value;
import com.chbachman.cminus.representation.value.Variable;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Created by Chandler on 4/14/17.
 */
public class ForStatement implements Control {

    Value minimum;
    Value maximum;

    Variable index;

    public ForStatement(CMinusParser.ForStatementContext ctx, Scope scope) {
        if (ctx.range() != null) {
            CMinusParser.RangeContext range = ctx.range();

            minimum = Value.parse(range.value(0), scope);
            maximum = Value.parse(range.value(1), scope);

            if (minimum.type() != Type.Native.INT.type || maximum.type() != Type.Native.INT.type) {
                throw new RuntimeException("For Loop without int range. " + range.getText());
            }

            index = new Variable(range.ID().getText(), Type.Native.INT.type);
        }
    }

    @Override
    public void setupScope(Scope scope) {
        scope.addVariable(index);
    }

    @Override
    public String first() {
        StringBuilder s = new StringBuilder("for (")
                .append("int ")
                .append(index.name)
                .append(" = ")
                .append(minimum.value())
                .append("; ")
                .append(index.name)
                .append(" <= ")
                .append(maximum.value())
                .append("; ")
                .append(index.name)
                .append("++) {");

        return s.toString();
    }

    @Override
    public String last() {
        return "}";
    }
}

File: src/main/java/com/chbachman/cminus/representation/control/ForStatement.kt
package com.chbachman.cminus.representation.control

import com.chbachman.cminus.gen.CMinusParser
import com.chbachman.cminus.representation.Parser
import com.chbachman.cminus.representation.Scope
import com.chbachman.cminus.representation.Type
import com.chbachman.cminus.representation.function.CodeBlockHolder
import com.chbachman.cminus.representation.value.Expression
import com.chbachman.cminus.representation.value.Identifier
import com.chbachman.cminus.representation.value.Statement
import com.chbachman.cminus.representation.value.Variable

/**
 * Created by Chandler on 4/14/17.
 * Handles the various types of For Loops
 */
class ForStatement(ctx: CMinusParser.ForStatementContext, scope: Scope) : CodeBlockHolder(), Control {

    private val minimum: Expression
    private val maximum: Expression

    private val index: Variable

    override val statements: List<Statement>

    init {
        val range = ctx.range()

        minimum = Parser.parse(range.value(0), scope)
        maximum = Parser.parse(range.value(1), scope)

        if (minimum.type != Type.Native.INT.type || maximum.type != Type.Native.INT.type) {
            throw RuntimeException("For Loop without int range. ${range.text}")
        }

        index = Variable(Identifier(range.ID().text), Type.Native.INT.type)

        scope.pushScope(this)
        statements = Parser.parse(ctx.codeBlock(), scope)
        scope.popScope()
    }

    override fun setupScope(scope: Scope) {
        scope.addVariable(index)
    }

    override val first: String
        get() {
            val i = index.name
            return "for (int $i = ${minimum.expression}; $i <= ${maximum.expression}; $i++) {"
        }

    override val last = "}"
}
